internal/app/go-filecoin/storage_market_connector/common.go

```diff
@@ -0,0 +1,290 @@
+
+	"github.com/filecoin-project/go-address"
+	smtypes "github.com/filecoin-project/go-fil-markets/shared/types"
+	"github.com/filecoin-project/go-fil-markets/storagemarket"
+	spaabi "github.com/filecoin-project/specs-actors/actors/abi"
+	spasm "github.com/filecoin-project/specs-actors/actors/builtin/storage_market"
+	spautil "github.com/filecoin-project/specs-actors/actors/util"
+	"github.com/ipfs/go-cid"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-filecoin/internal/app/go-filecoin/plumbing/msg"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/block"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/types"
+	fcsm "github.com/filecoin-project/go-filecoin/internal/pkg/vm/actor/builtin/storagemarket"
+	fcaddr "github.com/filecoin-project/go-filecoin/internal/pkg/vm/address"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/wallet"
+)
+
+type chainReader interface {
+	Head() block.TipSetKey
+	GetTipSet(block.TipSetKey) (block.TipSet, error)
+	GetActorStateAt(ctx context.Context, tipKey block.TipSetKey, addr fcaddr.Address, out interface{}) error
+}
+
+// Implements storagemarket.StateKey
+type stateKey struct {
+	ts     block.TipSetKey
+	height uint64
+}
+
+func (k *stateKey) Height() uint64 {
+	return k.height
+}
+
+// WorkerGetter is a function that can retrieve the miner worker for the given address from actor state
+type WorkerGetter func(ctx context.Context, minerAddr fcaddr.Address, baseKey block.TipSetKey) (fcaddr.Address, error)
+
+type connectorCommon struct {
+	chainStore   chainReader
+	waiter       *msg.Waiter
+	wallet       *wallet.Wallet
+	outbox       *message.Outbox
+	workerGetter WorkerGetter
+}
+
+// MostRecentStateId returns the state key from the current head of the chain.
+func (c *connectorCommon) MostRecentStateId(ctx context.Context) (storagemarket.StateKey, error) { // nolint: golint
+	key := c.chainStore.Head()
+	ts, err := c.chainStore.GetTipSet(key)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &stateKey{key, uint64(ts.At(0).Height)}, nil
+}
+
+func (c *connectorCommon) wait(ctx context.Context, mcid cid.Cid, pubErrCh chan error) (*types.MessageReceipt, error) {
+	receiptChan := make(chan *types.MessageReceipt)
+	errChan := make(chan error)
+
+	err := <-pubErrCh
+	if err != nil {
+		return nil, err
+	}
+
+	go func() {
+		err := c.waiter.Wait(ctx, mcid, func(b *block.Block, message *types.SignedMessage, r *types.MessageReceipt) error {
+			receiptChan <- r
+			return nil
+		})
+		if err != nil {
+			errChan <- err
+		}
+	}()
+
+	select {
+	case receipt := <-receiptChan:
+		if receipt.ExitCode != 0 {
+			return nil, xerrors.Errorf("non-zero exit code: %d", receipt.ExitCode)
+		}
+
+		return receipt, nil
+	case err := <-errChan:
+		return nil, err
+	case <-ctx.Done():
+		return nil, xerrors.New("context ended prematurely")
+	}
+}
+
+func (c *connectorCommon) addFunds(ctx context.Context, from address.Address, addr address.Address, amount tokenamount.TokenAmount) error {
```

Comment:
That's potentially true (note "addFunds" here actually means add storage collateral funds). This method simply puts a message on chain. It's not entirely clear if the CLI endpoint would go directly to this method or if it would interact with it indirectly through the storage miner. For that reason, it's probably best to wait before we try to share this code, but I agree in principle.